pkgs/by-name/nu/nuv/package.nix

```diff
@@ -0,0 +1,68 @@
+{ lib
+, stdenv
+, pkgs
+, callPackage
+, fetchFromGitHub
+, fetchurl
+, buildGoModule
+, makeWrapper
+, breakpointHook
+, jq
+, curl
+, kubectl
+, eksctl
+, kind
+, k3sup
+, coreutils
+}:
+
+let
+  branch = "3.0.0";
+  version = "3.0.1-beta.2405292059";
+  pname = "nuv";
+in
+buildGoModule {
+  inherit pname version;
+
+  src = fetchFromGitHub {
+    owner = "nuvolaris";
+    repo = "nuv";
+    rev = version;
+    hash = "sha256-MdnBvlA4S2Mi/bcbE+O02x+wvlIrsK1Zc0dySz4FB/w=";
+  };
+
+  subPackages = [ "." ];
+  vendorHash = "sha256-JkQbQ2NEaumXbAfsv0fNiQf/EwMs3SDLHvu7c/bU7fU=";
+
+  nativeBuildInputs = [ makeWrapper jq curl breakpointHook ];
+
+  buildInputs = [ kubectl eksctl kind k3sup coreutils ];
+
+  ldflags = [
+    "-X main.NuvVersion=${version}"
+    "-X main.NuvBranch=${branch}"
+  ];
+
+  # false because tests require some modifications inside nix-env
+  doCheck = false;
+
+  postInstall = ''
+    makeWrapper ${coreutils}/bin/coreutils $out/bin/coreutils
+    makeWrapper ${kubectl}/bin/kubectl $out/bin/kubectl
+    makeWrapper ${eksctl}/bin/eksctl $out/bin/eksctl
+    makeWrapper ${kind}/bin/kind $out/bin/kind
+    makeWrapper ${k3sup}/bin/k3sup $out/bin/k3sup
```

Comment:
Do you think it would be better to wrap the `nuv` executable with a wrapper that sets `NUV_BIN` to a `symlinkJoin` of these packages? Or is there a good reason you've done it in this specific way?

Reply:
@h7x4 nuv command will do that by himself. So it's not necessary another wrapper

Reply:
I'm not too worried about whether `nuv` picks it up or not. The thing that bothers me is the executable name clash. Let's say I was a user, and had constructed this environment:

```
pkgs.mkShell {
  packages = with pkgs; [
    nuv
    (kubectl.override {
      ... # misc overrides and patches to kubectl
    })
    ...
  ];
};
```

And inside of this environment, I run `kubectl`. I would expect my shell to run the overridden one, but AFAIK there is no good guarantee that the shell won't pick up the "old kubectl" from within `${nuv}/bin` first.

From what I'm reading in the upstream README, you can specify where the binary looks for executables with the `NUV_BIN` envvar, so I suggest we do something like this instead:

```nix
postInstall = let
  nuv-bin = pkgs.symlinkJoin {
    name = "nuv-bin";
    paths = [
      coreutils
      kubectl
      eksctl
      kind
      k3sup
    ];
  };
in ''
  makeWrapper $out/bin/nuv \
    --set NUV_BIN "${nuv-bin}/bin"
'';
```

This way, we can isolate `nuv`s executables, and not mess up the rest of the environment.

Reply:
@h7x4 thank you for insisting and for your observations: they are convincing and in the meantime I have studied the matter better. I'll test it with your changes and then we'll move on.

Reply:
Hello @h7x4 i tried your code inside the package.nix and it ended with an error:

```nix
Running phase: configurePhase
Running phase: buildPhase
Building subPackage ./.
Running phase: installPhase
dirname: unrecognized option '--set'
Try 'dirname --help' for more information.
mkdir: cannot create directory '': No such file or directory
/nix/store/1bmd2z777d232v0ikizk1h46vq881q0k-make-shell-wrapper-hook/nix-support/setup-hook: line 131: pop_var_context: head of shell_variables not a function context
build failed in installPhase with exit code 1
```

it seems that the makeWrapper command is missing something. Can you help me with that?

Reply:
Sorry, `wrapProgram` seems more suitable. It's a version of `makeWrapper` that automatically renames the unwrapped executable inplace.

Reply:
@h7x4 check if my last commit is ok for you. I need still some test.

Reply:
The code looks nice, the derivation builds. I have no idea how to use the tool though, so it would be nice if you could verify that it still runs as expected, and squash the changes into `nuv: init at 3.0.1-beta.2405292059`

Reply:
I agree: I will do some extra test and let you know.

Reply:
Hello @h7x4. I did my tests and here is the outcome.

I've builded a shell with the following `default.nix`: 

```nix
{ pkgs ? import <nixpkgs> {}, 
  version ? ""
}:
let  
  nuv = pkgs.callPackage ./nuv.nix {}; 
in
pkgs.mkShell {
  buildInputs = [
    nuv
    pkgs.which
    pkgs.curl
    pkgs.openssh
  ];
}
```
Curl and openssh are requirements for nuv to run.
So i've which and nuv commands. Started it with --pure.

```bash
bruno:tmp.BwyxQT05Mi/ $ nix-shell --pure default.nix                                                                  [9:19:41]

[nix-shell:/run/user/1000/tmp.BwyxQT05Mi]$ which nuv
/nix/store/8f22dwscgdsdji3sv1i5n7ap52mkwg9v-nuv-3.0.1-beta.2405062139/bin/nuv

[nix-shell:/run/user/1000/tmp.BwyxQT05Mi]$ strings /nix/store/8f22dwscgdsdji3sv1i5n7ap52mkwg9v-nuv-3.0.1-beta.2405062139/bin/nuv
#! /nix/store/306znyj77fv49kwnkpxmb0j2znqpa8bj-bash-5.2p26/bin/bash -e
export NUV_BIN='/nix/store/96rn7wbpa5gv85wlxbb3wz4assbznp8m-nuv-bin/bin'
exec -a "$0" "/nix/store/8f22dwscgdsdji3sv1i5n7ap52mkwg9v-nuv-3.0.1-beta.2405062139/bin/.nuv-wrapped"  "$@" 
```

The nuv command is wrapped and the NUV_BIN path have been set: all fine so far :)

Now if i list the content of the NUV_BIN folder there are all the symlinks we expect:

```bash
[nix-shell:/run/user/1000/tmp.BwyxQT05Mi]$ ls -lah /nix/store/96rn7wbpa5gv85wlxbb3wz4assbznp8m-nuv-bin/bin | grep nix
lrwxrwxrwx 1 root root   75 Jan  1  1970 .k3sup-wrapped -> /nix/store/c32c2fri7wgjv06iv5kydv42nl4ili41-k3sup-0.13.5/bin/.k3sup-wrapped
lrwxrwxrwx 1 root root   71 Jan  1  1970 coreutils -> /nix/store/php4qidg2bxzmm79vpri025bqi0fa889-coreutils-9.5/bin/coreutils
lrwxrwxrwx 1 root root   69 Jan  1  1970 eksctl -> /nix/store/y67k1mj8fh7i4x2ra0dl6i38jazqyi1d-eksctl-0.179.0/bin/eksctl
lrwxrwxrwx 1 root root   66 Jan  1  1970 k3sup -> /nix/store/c32c2fri7wgjv06iv5kydv42nl4ili41-k3sup-0.13.5/bin/k3sup
lrwxrwxrwx 1 root root   64 Jan  1  1970 kind -> /nix/store/8if8dwbvyalf0cfk0vkaysvcmkva9syh-kind-0.23.0/bin/kind
lrwxrwxrwx 1 root root   70 Jan  1  1970 kubectl -> /nix/store/05k95p72niq1gh39gv1g9n0ivsgl4hiy-kubectl-1.30.1/bin/kubectl
```

The nuv command seems to find kubectl and the other commands in it's NUV_BIN path:

```bash
[nix-shell:/run/user/1000/tmp.BwyxQT05Mi]$ nuv util kubectl help
kubectl controls the Kubernetes cluster manager.

 Find more information at: https://kubernetes.io/docs/reference/kubectl/
[...]
```

It seems to work correctly. I need to ask @sciabarracom if there are other implications with this before proceed with the solution.

Reply:
Hello @h7x4. It's ok for us to proceed. Let me know if you need to change something more.

Reply:
Good to hear that it works, just squash the changes into `nuv: init at 3.0.1-beta.2405292059`, and I can merge it